Backend/src/model/course.py

```diff
@@ -1,13 +1,93 @@
+from __future__ import annotations
 from dataclasses import dataclass
-from typing import List
+from typing import Callable, List
 from .section import Section
+from .date import Date
+from .day_of_week import DayOfWeek
+from .filter import Filter
 
 @dataclass
 class Course:
+    code: str
     name: str
-    title: str
     term: str
     prerequisite: str
-    lectureSections: List[Section]
-    labSections: List[Section]
+    lecture_sections: List[Section]
+    lab_sections: List[Section]
+    section_id_filter: str
+
+    @staticmethod
+    def filter_all_courses(courses: List[Course], filter: Filter):
+        if filter is None:
+            raise ValueError("Parameter 'filter' cannot be None!")
+
+        for course in courses:
+            course.filter(filter)
+
+    @staticmethod
+    def filter_sections(sections: List[Section], compare_function: Callable[[Date], bool]) -> None:
+        i = 0
+        while i < len(sections):
+            removed_section = False
+            for date in sections[i].times:
+                if compare_function(date):
+                    del sections[i]
+                    removed_section = True
+                    break
+                
+            if not removed_section:
+                i += 1
+
+    def filter(self, filter: Filter) -> None:
+        if filter.before_time:
+            self.filter_before_time(filter.before_time)
+        
+        if filter.after_time:
+            self.filter_after_time(filter.after_time)
+
+        if filter.day_of_week:
+            self.filter_day_off(filter.day_of_week)
+        
+        if self.section_id_filter:
+            self.filter_by_section()
+
+    def filter_before_time(self, time: str) -> None:
+        '''
+        Filter out all sections that take place before filter time
+        '''
+        
+        def compare_function(date: Date):
+            return date.get_start_time_as_float() < Date.convert_time_to_float(time)
+
+        Course.filter_sections(self.lecture_sections, compare_function)
+        Course.filter_sections(self.lab_sections, compare_function)       
+
+
+
+    def filter_after_time(self, time: str) -> None:
+        '''
+        Filter out all sections that take place after filter time
+        '''
+        def compare_function(date: Date):
+            return date.get_end_time_as_float() > Date.convert_time_to_float(time)
+
+        Course.filter_sections(self.lecture_sections, compare_function)
+        Course.filter_sections(self.lab_sections, compare_function) 
+                
+
+    def filter_day_off(self, day_of_week: DayOfWeek) -> None:
```

Comment:
Are there future plans to include a filter for "most days off" (assuming this is feasible)?

Reply:
I guess this could be possible but this would change how we schedule courses not how we filter them.